High memory use (jetsam) with scaled-down tiled layers

smfr opened the pull request:
#### ae4b128c8c00089c79fed1e17b80e6279d7fa7a7
<pre>
High memory use (jetsam) with scaled-down tiled layers
<a href="https://bugs.webkit.org/show_bug.cgi?id=242675">https://bugs.webkit.org/show_bug.cgi?id=242675</a>
&lt;rdar://96303961&gt;

Reviewed by Tim Horton.

The site in question had some large composting layers with a scale(0.05) transform; this
triggers the creation of many visible tiles, causing jetsam. This got worse after
249828@main which triggered compositing for some negative z-index cases.

To fix this, impose a per-layer memory limit of 156MB (this is enough to cover a large display
with a 4-byte per pixel texture format) for tiled layers. For layers above this limit, adjust
the contents scale with factor between 0.05 and 1 which reduce the contents scale; this
may cause some blurriness, but that&apos;s preferable to triggering jetsam.

setContentsScaleLimitingFactor() is called in two places; first, when we swap to/from
tiled layers, and second when the exposed area of a tiled layer changes.

* LayoutTests/compositing/tiling/tiled-layers-limit-contents-scale-expected.txt: Added.
* LayoutTests/compositing/tiling/tiled-layers-limit-contents-scale.html: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::adjustContentsScaleLimitingFactor):
(WebCore::GraphicsLayerCA::setContentsScaleLimitingFactor):
(WebCore::GraphicsLayerCA::setVisibleAndCoverageRects):
(WebCore::GraphicsLayerCA::updateContentsScale):
(WebCore::GraphicsLayerCA::dumpAdditionalProperties const):
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setContentsScale):
(WebCore::TileController::adjustTileCoverageForDesktopPageScrolling const):
(WebCore::TileController::didRevalidateTiles):

Canonical link: <a href="https://commits.webkit.org/252437@main">https://commits.webkit.org/252437@main</a>
</pre>

mattwoodrow commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -1559,6 +1562,44 @@ TransformationMatrix GraphicsLayerCA::transformByApplyingAnchorPoint(const Trans
+    auto tileCoverageRect = intersection(m_coverageRect, bounds);
+    if (!tileCoverageRect.isEmpty()) {
+        double scaleFactor = deviceScaleFactor() * pageScaleFactor();
+        double memoryEstimate = tileCoverageRect.area() * scaleFactor * scaleFactor * 4.0;
```
4.0 here is the bytes per pixel, right? I'd prefer to use a named static const var for this, or possible actually read the runtime value from the layer.

I think we can sometimes be using higher than 4bpp, since we have the 10bit color format for P3 (RGB10_XR w30r)

smfr replied:
True, but there's some extra plumbing required to get that out for tiled backing. For now I'll hardcode 4 (this is only an estimate, anyway).

mattwoodrow commented on Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
```diff
@@ -101,6 +101,9 @@ static const unsigned cMaxLayerTreeDepth = 128;
 static const unsigned cMaxTotalBackdropFilterArea = 1242 * 2208 * 10;
 
+// Don't let a single tiled layer use more than 128MB of memory.
+static const unsigned cMaxScaledTiledLayerMemorySize = 1024 * 1024 * 128;
```
It might be ok to go over larger here, so that we're only catching the most egregious cases.

Whatever value we chose is going to be a tradeoff between making layers pixelated and jetsams, so leaning towards the existing behaviour (always jetsam) and just fixing the known bad cases might be enough.

As mentioned below, we're using 1536x1536 8bpp tiles on iOS, so that's roughly ~9mb per tile, and there can be 6 visible at a time without scaling applied.

smfr replied:
It's unusual to reveal so many tiles on a tiled layers that we hit the contents scale adjustment. The two cases are something large scaled down with transforms, and something rotated in Y or X. When scaled down, the contents scale adjustment compensates in a way that usually still makes the tiles rendered at higher resolution than necessary, so blurriness will be rare, I think. I chose 128MB because the Apple Pro XDR display is 6016 * 3384 which is about 78MB.

smfr replied:
I chose 156MB for this patch, based on wanting 12 3x tiles on an iPhone display.

hortont424 commented on Source/WebCore/platform/graphics/ca/TileController.cpp:
```diff
@@ -103,14 +103,16 @@ void TileController::setNeedsDisplayInRect(const IntRect& rect)
     // The scale we get is the product of the page scale factor and device scale factor.
     // Divide by the device scale factor so we'll get the page scale factor.
-    scale /= deviceScaleFactor;
+    float scale = contentsScale/ deviceScaleFactor;
```
missing a space before the /

smfr committed:
High memory use (jetsam) with scaled-down tiled layers
https://bugs.webkit.org/show_bug.cgi?id=242675
<rdar://96303961>

Reviewed by Tim Horton.

The site in question had some large composting layers with a scale(0.05) transform; this
triggers the creation of many visible tiles, causing jetsam. This got worse after
249828@main which triggered compositing for some negative z-index cases.

To fix this, impose a per-layer memory limit of 156MB (this is enough to cover a large display
with a 4-byte per pixel texture format) for tiled layers. For layers above this limit, adjust
the contents scale with factor between 0.05 and 1 which reduce the contents scale; this
may cause some blurriness, but that's preferable to triggering jetsam.

setContentsScaleLimitingFactor() is called in two places; first, when we swap to/from
tiled layers, and second when the exposed area of a tiled layer changes.

* LayoutTests/compositing/tiling/tiled-layers-limit-contents-scale-expected.txt: Added.
* LayoutTests/compositing/tiling/tiled-layers-limit-contents-scale.html: Added.
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.cpp:
(WebCore::GraphicsLayerCA::adjustContentsScaleLimitingFactor):
(WebCore::GraphicsLayerCA::setContentsScaleLimitingFactor):
(WebCore::GraphicsLayerCA::setVisibleAndCoverageRects):
(WebCore::GraphicsLayerCA::updateContentsScale):
(WebCore::GraphicsLayerCA::dumpAdditionalProperties const):
(WebCore::GraphicsLayerCA::changeLayerTypeTo):
* Source/WebCore/platform/graphics/ca/GraphicsLayerCA.h:
* Source/WebCore/platform/graphics/ca/TileController.cpp:
(WebCore::TileController::setContentsScale):
(WebCore::TileController::adjustTileCoverageForDesktopPageScrolling const):
(WebCore::TileController::didRevalidateTiles):

Canonical link: https://commits.webkit.org/252437@main
